src/main/java/jenkins/bouncycastle/api/BouncyCastlePlugin.java

```diff
@@ -0,0 +1,89 @@
+package jenkins.bouncycastle.api;
+
+import java.io.File;
+import java.security.Provider;
+import java.security.Security;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.accmod.restrictions.suppressions.SuppressRestrictedWarnings;
+import hudson.Main;
+import hudson.Plugin;
+import hudson.PluginWrapper;
+import jenkins.model.Jenkins;
+import jenkins.util.AntClassLoader;
+
+@SuppressWarnings("deprecation") // there is no other way to achieve this at the correct lifecycle point
+@Restricted(NoExternalUse.class) // just for Jenkins access not part of the API
+public class BouncyCastlePlugin extends Plugin {
+
+    private static final Logger LOG = Logger.getLogger(BouncyCastlePlugin.class.getName());
+
+    private static final boolean isActive;
+
+    static {
+        // BouncyCastle FIPS is installed in the JVM, we can not install over the top of it so do not try
+        Provider p = Security.getProvider("BCFIPS");
+        isActive = (p == null);
+
+        LOG.log(Level.CONFIG,
+                isActive ? "BouncyCastle Providers from BouncyCastle API plugin will be active" :
+                           "Detected the presence of the BouncyCastle FIPS provider, the regular BouncyCastle JARs will not be available.");
+    }
+
+    @Override
+    @SuppressRestrictedWarnings(jenkins.util.AntClassLoader.class) // we are messing with the classloader and it has not changed in many many years
+    public void start() throws Exception {
+        if (!isActive) {
+            // Alternative BouncyCastle is installed do no not insert these libraries
+            return;
+        }
+        // this is the hairy part.
+        // add the BouncyCastle APIs into the classpath for other plugins (and this plugin to use!)
+        /*
+         * Whilst plugins that have code may go boom before this with class loading issues, extensions (at this point)
+         * have not been discovered, so this would only affect people using the deprecated `Plugin` class (like we are!)
+         */
+
+        final File optionalLibDir = getOptionalLibDirectory();
+        File[] optionalLibs = optionalLibDir.listFiles();
```

Comment:
nit: Maybe better to use NIO APIs here so you get exceptions with relevant error messages if something goes wrong.

Reply:
I agree in principal, but if this fails I think Jenkins probably has much bigger concerns (the unpacking of the hpi or war has failed, and as such Jenkins as a whole will likely have issues rather than just this plugin - or the plugin itself is borked and will not start correctly)